pkg/controller/disruption/disruption.go

```diff
@@ -173,94 +175,119 @@ func (dc *DisruptionController) finders() []podControllerFinder {
+	controllerKindDep = v1beta1.SchemeGroupVersion.WithKind("Deployment")
+)
+
 // getPodReplicaSets finds replicasets which have no matching deployments.
 func (dc *DisruptionController) getPodReplicaSets(pod *v1.Pod) ([]controllerAndScale, error) {
-	cas := []controllerAndScale{}
-	rss, err := dc.rsLister.GetPodReplicaSets(pod)
-	// GetPodReplicaSets returns an error only if no ReplicaSets are found.  We
-	// don't return that as an error to the caller.
+	var casSlice []controllerAndScale
+	controllerRef := controller.GetControllerOf(pod)
+	if controllerRef == nil {
+		return nil, nil
+	}
+	if controllerRef.Kind != controllerKindRS.Kind {
+		return nil, nil
+	}
+	rs, err := dc.rsLister.ReplicaSets(pod.Namespace).Get(controllerRef.Name)
 	if err != nil {
-		return cas, nil
-	}
-	controllerScale := map[types.UID]int32{}
-	for _, rs := range rss {
-		// GetDeploymentsForReplicaSet returns an error only if no matching
-		// deployments are found.
-		_, err := dc.dLister.GetDeploymentsForReplicaSet(rs)
-		if err == nil { // A deployment was found, so this finder will not count this RS.
-			continue
-		}
-		controllerScale[rs.UID] = *(rs.Spec.Replicas)
+		// The only possible error is NotFound, which is ok here.
+		return nil, nil
 	}
-
-	for uid, scale := range controllerScale {
-		cas = append(cas, controllerAndScale{UID: uid, scale: scale})
+	if rs.UID != controllerRef.UID {
+		return nil, nil
 	}
-
-	return cas, nil
+	controllerRef = controller.GetControllerOf(rs)
+	if controllerRef != nil && controllerRef.Kind == controllerKindDep.Kind {
+		// Skip RS if it's controlled by a Deployment.
+		return nil, nil
+	}
+	casSlice = append(casSlice, controllerAndScale{rs.UID, *(rs.Spec.Replicas)})
+	return casSlice, nil
 }
 
 // getPodStatefulSet returns the statefulset managing the given pod.
 func (dc *DisruptionController) getPodStatefulSets(pod *v1.Pod) ([]controllerAndScale, error) {
-	cas := []controllerAndScale{}
-	ss, err := dc.ssLister.GetPodStatefulSets(pod)
-
-	// GetPodStatefulSets returns an error only if no StatefulSets are found. We
-	// don't return that as an error to the caller.
-	if err != nil {
-		return cas, nil
+	var casSlice []controllerAndScale
+	controllerRef := controller.GetControllerOf(pod)
+	if controllerRef == nil {
+		return nil, nil
 	}
-
-	controllerScale := map[types.UID]int32{}
-	for _, s := range ss {
-		controllerScale[s.UID] = *(s.Spec.Replicas)
+	if controllerRef.Kind != controllerKindSS.Kind {
+		return nil, nil
 	}
-
-	for uid, scale := range controllerScale {
-		cas = append(cas, controllerAndScale{UID: uid, scale: scale})
+	ss, err := dc.ssLister.StatefulSets(pod.Namespace).Get(controllerRef.Name)
+	if err != nil {
+		// The only possible error is NotFound, which is ok here.
+		return nil, nil
+	}
+	if ss.UID != controllerRef.UID {
+		return nil, nil
 	}
 
-	return cas, nil
+	casSlice = append(casSlice, controllerAndScale{ss.UID, *(ss.Spec.Replicas)})
```

Comment:
@kargakis i was hoping to do that change in a subsequent PR, since that would also involve modifying the callers and understanding that logic. Is that ok ?